src/coprocessor/statistics/fmsketch.rs

```diff
@@ -0,0 +1,164 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// FIXME: remove following later
+#![allow(dead_code)]
+
+use std::collections::HashSet;
+use byteorder::{ByteOrder, LittleEndian};
+use murmur3::murmur3_x64_128;
+use tipb::analyze;
+
+/// `FMSketch` is used to count the approximate number
+/// of distinct elements in a `[Datum]`.
+pub struct FMSketch {
+    mask: u64,
+    max_size: usize,
+    hash_set: HashSet<u64>,
+}
+
+impl FMSketch {
+    pub fn new(max_size: usize) -> FMSketch {
+        FMSketch {
+            mask: 0,
+            max_size: max_size,
+            hash_set: HashSet::new(),
+        }
+    }
+
+    // ndv returns the approximate number of distinct elements
+    pub fn ndv(&self) -> u64 {
+        (self.mask + 1) as u64 * (self.hash_set.len() as u64)
+    }
+
+    pub fn insert(&mut self, mut bytes: &[u8]) {
+        let hash = {
+            let mut out: [u8; 16] = [0; 16];
+            murmur3_x64_128(&mut bytes, 0, &mut out);
+            LittleEndian::read_u64(&out[0..8])
```

Comment:
Why initialized as 16 but only takes 8?

Reply:
Usually, the 64-bit version of murmur is done by drop the second half of the 128-bit version, so it will only takes 8.